Allow a list of files to be checked

sammarshallou opened the pull request:
Here's a simple change that lets you paste in a list of files to the textarea. It's super useful if checking files manually, because you can just do 'git show --name-only' on your commit, and then paste in the list of files to check.

I also increased the time limit to 600 because checking is so slow nowadays (which actually, is why it's often more useful to check a list of files rather than a whole folder at once - although often you also want to check files from multiple folders).

stronk7 commented on index.php:
```diff
@@ -43,27 +43,33 @@
+    foreach ($paths as $path) {
+        $fullpath = $CFG->dirroot . '/' . trim($path, '/');
+        if (!is_file($fullpath) && !is_dir($fullpath)) {
+            $fullpaths = [];
```
Why an incorrect entry does reset all full-paths? Sure I'm missing something, hence asking.

sammarshallou replied:
The reason for clearing the array is so that the code underneath (that actually does the checking) does not run. I thought it was better to give an error if a path was invalid, rather than still processing all the other ones. I'll add a comment to explain this & also check the behaviour actually works.

sammarshallou replied:
(as mentioned in comment below, I actually changed the way this bit works instead, hopefully it makes more sense now.)

stronk7 commented on index.php:
```diff
@@ -43,27 +43,33 @@
+        $fullpath = $CFG->dirroot . '/' . trim($path, '/');
+        if (!is_file($fullpath) && !is_dir($fullpath)) {
+            $fullpaths = [];
+            break;
```
More yet, and related with the above, why break? Wouldn't be better to just accumulate the correct ones and skip the non-existing ones?

stronk7 commented on index.php:
```diff
@@ -26,13 +26,13 @@
 require_once($CFG->dirroot . '/local/codechecker/locallib.php');
 
-$path = optional_param('path', '', PARAM_PATH);
+$pathlist = optional_param('path', '', PARAM_RAW);
```
While it's not the end of the world... maybe we should apply for PARAM_PATH validation @ some place below? I was expecting `local_codechecker_clean_path()` to, maybe do that, but it's not, [right now](https://github.com/moodlehq/moodle-local_codechecker/blob/24ee399b2439e4ada87cceff046733b0563386de/locallib.php#L294).

Maybe within the `$paths` loop, first thing...

sammarshallou replied:
Yes good point, I'll add it!

sammarshallou commented:
I've pushed a new version and changed the way it works a bit - now, it will output the failures [file not found] for each path (instead of stopping at the first path that fails). So you get multiple errors if necessary. If there are any failures, it doesn't run the checks. I tried this out, it seems to work OK.

Also added in the clean_param as pointed out.

I think this one is better?

stronk7 reviewed (approved):
Looks perfect, thanks!

stronk7 commented:
Fun, just real usage experience... I was running some tests in the UI against mod/label... and had to kill apache a number of times... just to discover that, with the new text box, if you enter some blank line... the whole dirroot is checked (and it can take hours).

And it's really easy for that to happen (if you were used to the old input field and form being submitted on return). In fact I had 3 in my box, heh.

So we need to add something like this, to prevent blank lines to be computed as dirroot:

```
@@ -61,7 +72,10 @@ if ($pathlist) {
     $failed = false;
     $fullpaths = [];
     foreach ($paths as $path) {
-        $path = clean_param($path, PARAM_PATH);
+        $path = trim(clean_param($path, PARAM_PATH));
+        if (empty($path)) {
+            continue;
+        }
         $fullpath = $CFG->dirroot . '/' . trim($path, '/');
         if (!is_file($fullpath) && !is_dir($fullpath)) {
             echo $output->invald_path_message($path);
```

Will create a patch soon with that...
